apps/avifgainmaputil/convert_command.cc

```diff
@@ -0,0 +1,108 @@
+  argparse_.add_argument(arg_output_filename_, "output_image.avif");
+  argparse_.add_argument(arg_swap_base_, "--swap-base")
+      .help("Make the HDR image the base image")
+      .action(argparse::Action::STORE_TRUE)
+      .default_value("false");
+  argparse_.add_argument(arg_gain_map_quality_, "--qgain-map")
+      .help("Quality for the gain map (0-100, where 100 is lossless)")
+      .default_value("60");
+  argparse_.add_argument<CicpValues, CicpConverter>(arg_cicp_, "--cicp")
+      .help(
+          "Set the cicp values for the input image, expressed as "
+          "P/T/M where P = color primaries, T = transfer characteristics, "
+          "M = matrix coefficients.");
+  arg_image_encode_.Init(argparse_, /*can_have_alpha=*/false);
+  arg_image_read_.Init(argparse_);
+}
+
+avifResult ConvertCommand::Run() {
+  const avifPixelFormat pixel_format =
+      static_cast<avifPixelFormat>(arg_image_read_.pixel_format.value());
+
+  ImagePtr image(avifImageCreateEmpty());
+  if (image == nullptr) {
+    return AVIF_RESULT_OUT_OF_MEMORY;
+  }
+
+  const avifAppFileFormat file_format = avifReadImage(
+      arg_input_filename_.value().c_str(), pixel_format, arg_image_read_.depth,
+      AVIF_CHROMA_DOWNSAMPLING_AUTOMATIC, arg_image_read_.ignore_profile,
+      /*ignoreExif=*/false,
+      /*ignoreXMP=*/false,
+      /*allowChangingCicp=*/true,
+      /*ignoreGainMap=*/false, image.get(),
+      /*outDepth=*/nullptr,
+      /*sourceTiming=*/nullptr,
+      /*frameIter=*/nullptr);
+  if (file_format == AVIF_APP_FILE_FORMAT_UNKNOWN) {
+    std::cout << "Failed to decode image: " << arg_input_filename_;
+    return AVIF_RESULT_INVALID_ARGUMENT;
+  }
+  if (arg_cicp_.provenance() == argparse::Provenance::SPECIFIED) {
+    image->colorPrimaries = arg_cicp_.value().color_primaries;
+    image->transferCharacteristics = arg_cicp_.value().transfer_characteristics;
+    image->matrixCoefficients = arg_cicp_.value().matrix_coefficients;
+  } else if (image->icc.size == 0 &&
+             image->colorPrimaries == AVIF_COLOR_PRIMARIES_UNSPECIFIED &&
+             image->transferCharacteristics ==
+                 AVIF_COLOR_PRIMARIES_UNSPECIFIED) {
+    // If there is no ICC and no CICP, assume sRGB by default.
+    image->colorPrimaries = AVIF_COLOR_PRIMARIES_BT709;
+    image->transferCharacteristics = AVIF_TRANSFER_CHARACTERISTICS_SRGB;
+  }
+
+  if (image->gainMap.image == nullptr) {
+    std::cerr << "Input image " << arg_input_filename_
+              << " does not contain a gain map\n";
+    return AVIF_RESULT_INVALID_ARGUMENT;
+  }
+
+  if (arg_swap_base_) {
+    int depth = arg_image_read_.depth;
+    if (depth == 0) {
+      depth = image->gainMap.metadata.alternateHdrHeadroomN == 0 ? 8 : 10;
+    }
+    ImagePtr new_base(
+        avifImageCreate(image->width, image->height, depth, image->yuvFormat));
+    const avifResult result = ChangeBase(std::move(*image), new_base.get());
```

Comment:
`image` contains fields (such as `yuvPlane`) that will be freed by `avifImageDestroy()`, leading to a use-after-move error when the dtor of `ImagePtr` is called.  
Either change the signature to `ChangeBase(ImagePtr&& image` or to `ChangeBase(avifImage& image`.

Reply:
Yeah ok I was trying ot be clever but the memory management is indeed a bit iffy especially when the image can come either from a unique_ptr or an avifDecoder